crates/miden-objects/src/asset/token_symbol.rs

```diff
@@ -26,99 +50,181 @@ impl From<TokenSymbol> for Felt {
-                "token symbol value {} cannot exceed {}",
-                felt.as_int(),
-                TokenSymbol::MAX_ENCODED_VALUE
-            )));
+        if felt.as_int() > Self::MAX_ENCODED_VALUE {
+            return Err(TokenSymbolError::ValueTooLarge(felt.as_int()));
         }
         Ok(TokenSymbol(felt))
     }
 }
 
 // HELPER FUNCTIONS
 // ================================================================================================
-// Utils to encode and decode the token symbol as a Felt. Token Symbols can consists of up to 6
-// characters , e.g., A = 0, ...
-fn encode_symbol_to_felt(s: &str) -> Result<Felt, AssetError> {
+
+/// Encodes the provided token symbol string into a single [`Felt`] value.
+///
+/// The alphabet used in the decoding process consists of the Latin capital letters as defined in
+/// the ASCII table, having the length of 26 characters.
+///
+/// The encoding is performed by multiplying the intermediate encrypted value by the length of the
+/// used alphabet and adding the relative index of the character to it. At the end of the encoding
+/// process the length of the initial token string is added to the encrypted value.
+///
+/// Relative character index is computed by subtracting the index of the character "A" (65) from the
+/// index of the currently processing character, e.g., `A = 65 - 65 = 0`, `B = 66 - 65 = 1`, `...` ,
+/// `Z = 90 - 65 = 25`.
+///
+/// # Errors
+/// Returns an error if:
+/// - The length of the provided string is less than 1 or greater than 6.
+/// - The provided token string contains characters that are not uppercase ASCII.
+fn encode_symbol_to_felt(s: &str) -> Result<Felt, TokenSymbolError> {
     if s.is_empty() || s.len() > TokenSymbol::MAX_SYMBOL_LENGTH {
-        return Err(AssetError::TokenSymbolError(format!(
-            "token symbol of length {} is not between 1 and 6 characters long",
-            s.len()
-        )));
+        return Err(TokenSymbolError::InvalidLength(s.len()));
     } else if s.chars().any(|c| !c.is_ascii_uppercase()) {
-        return Err(AssetError::TokenSymbolError(format!(
-            "token symbol {} contains characters that are not uppercase ASCII",
-            s
-        )));
+        return Err(TokenSymbolError::InvalidCharacter(s.to_string()));
     }
 
     let mut encoded_value = 0;
     for char in s.chars() {
         let digit = char as u64 - b'A' as u64;
-        assert!(digit < 26);
-        encoded_value = encoded_value * 26 + digit;
+        debug_assert!(digit < TokenSymbol::ALPHABET_LENGTH);
+        encoded_value = encoded_value * TokenSymbol::ALPHABET_LENGTH + digit;
     }
 
+    // add token length to the encoded value to be able to decode the exact number of characters
+    encoded_value = encoded_value * TokenSymbol::ALPHABET_LENGTH + s.len() as u64;
+
     Ok(Felt::new(encoded_value))
 }
 
-fn decode_felt_to_symbol(encoded_felt: Felt) -> String {
+/// Decodes the [`TokenSymbol`] into a token symbol string.
```

Comment:
I think we are decoding a felt representation of the token symbol (rather than `TokenSymbol` itself).